pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1343,97 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
+	var localPullSecret string
+	var err error
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	if cd.Spec.PullSecret != nil {
+		localPullSecret, err = controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return "", err
+			}
+		}
+	}
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", errors.Wrap(err, errMsg)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", errors.New(errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status.
+// It returns true when cluster deployment status has been updated.
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+	var err error
+	pullSecretObjExists := true
+	existingPullSecretObj := &corev1.Secret{}
+	mergedSecretName := constants.MergedPullSecretName
+	err = r.Get(context.TODO(), types.NamespacedName{Name: mergedSecretName, Namespace: cd.Namespace}, existingPullSecretObj)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cdLog.Info("Existing pull secret object not found")
+			pullSecretObjExists = false
+		} else {
+			return false, errors.Wrap(err, "Error getting pull secret from cluster deployment")
+		}
+	}
+
+	// create a new pull secret object
+	newPullSecretObj := generatePullSecretObj(
+		pullSecret,
+		mergedSecretName,
+		cd,
+	)
+	if pullSecretObjExists {
+		existingPullSecret, ok := existingPullSecretObj.Data[corev1.DockerConfigJsonKey]
+		if !ok {
+			return false, errors.New(fmt.Sprintf("Pull secret %s did not contain key %s", mergedSecretName, corev1.DockerConfigJsonKey))
+		}
+		if controllerutils.GetHashOfPullSecret(string(existingPullSecret)) == controllerutils.GetHashOfPullSecret(pullSecret) {
+			cdLog.Info("Existing and the new merged pull secret are same")
+			return false, nil
+		}
+		cdLog.Info("Existing merged pull secret hash did not match with latest merged pull secret")
+		err = r.Update(context.TODO(), newPullSecretObj)
```

Comment:
This update will always fail with a conflict. You need to update the existing pull secret instead of the new pull secret.

Reply:
I did not get you. Because I use the same merged secret name. Also it is not failing in my testing.

Reply:
Every kube resource has a revision. When you update a resource, you must pass the revision back to the API server. If the revision supplied does not match the revision that the API server knows about, then the API server will reject the request as a conflict. This is to ensure that one update does not conflict with another concurrent update.

In this case, you are passing to the update a new object which does not have a revision. The API server will reject that request. Has this been passing using a real client connected to a real API server or just in unit tests?